runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt

```diff
@@ -0,0 +1,167 @@
+package kotlinx.serialization
+
+import kotlinx.serialization.internal.StringDescriptor
+import kotlinx.serialization.internal.nullable
+import kotlinx.serialization.json.internal.BEGIN_LIST
+import kotlinx.serialization.json.internal.BEGIN_OBJ
+import kotlinx.serialization.json.internal.END_LIST
+import kotlinx.serialization.json.internal.END_OBJ
+import kotlinx.serialization.modules.EmptyModule
+import kotlinx.serialization.modules.SerialModule
+import kotlinx.serialization.modules.getContextualOrDefault
+
+
+/**
+ * @param encodeDefaults if false we will always write nulls explicitly.
+ *          If true we will omit the property if it has a default value.
+ */
+class DynamicObjectSerializer(private val encodeDefaults: Boolean = false, val context: SerialModule = EmptyModule) {
+
+    fun <T> serialize(strategy: SerializationStrategy<T>, obj: T): dynamic {
+
+        // handle singular primitive and enum values immediately
+        if (strategy.descriptor.kind is PrimitiveKind) {
+            if (obj is Char) {
+                val value: Char = obj // without the explicit type annotation int value would be used
+                return value.toString()
+            }
+            if (obj is Long) {
+                return obj.toDouble()
+            }
+            return obj
+        }
+        if (strategy.descriptor.kind is UnionKind.ENUM_KIND) {
+            return obj.unsafeCast<Enum<*>>().name
```

Comment:
Does this still work if we have a custom serializer for the enum that _doesn't_ use `name`?

Reply:
The if block will match any singular enum e.g. `DynamicObjectSerializer.serialize(Color.RED)`.

I am not sure what you are getting at. Are you worried about about a use-case where someone would want their enums to be serialised to a different name?

```
enum class Color {
@SerialName("red")
RED
}
```

The code above would not consider the SerialName annotation, or a custom serialiser if there were one. I think we would have to introduce a `SingularValueEncoder` to handle the case correctly.

Reply:
The if block will match any singular enum e.g. `DynamicObjectSerializer.serialize(Color.RED)`.

I am not sure what you are getting at. Are you worried about about a use-case where someone would want their enums to be serialised to a different name?

```
enum class Color {
@SerialName("red")
RED
}
```

The code above would not consider the SerialName annotation, or a custom serialiser if there were one. I think we would have to introduce a separate `DynamicValueEncoder` class to handle the case correctly, because `DynamicObjectEncoder` only deals with classes and collections.

Reply:
> Are you worried about about a use-case where someone would want their enums to be serialised to a different name?

Yeah that's exactly it. 

I know several folks (including myself) who use `@SerialName`, `@SerialId`, or custom `Serializer` implementations for enum classes (and of course other classes) since `ordinal`/`name` might not be consistent across codebases that still need to communicate with each other.

Reply:
It would be nice to support enums with `@SerialName` here despite singular customized enums are rare case —because it is already supported in `encodeEnum` for wrapped enums. I've not tested it myself (yet), but i believe that replacing this line with `strategy.descriptor.getElementName(obj.unsafeCast<Enum<*>>().ordinal)` would be sufficient — since custom serial names are stored in descriptor.